src/server.rs

```diff
@@ -128,27 +163,59 @@ impl ServerState {
             RequestAction::KeepAlive       => unimplemented!(),
             RequestAction::ListPlayers     => {
                 let mut players = vec![];
-                let player: &Player = self.players.get(player_id.0).unwrap();
-                if player.game == None {
+                let player: &Player = self.players.get(&player_id).unwrap();
+                if player.game_info == None {
                     return ResponseCode::BadRequest(Some("cannot list players because in lobby.".to_owned()));
                 };
-                let game_slot_id = &player.game.as_ref().unwrap().game_slot_id;  // unwrap ok because of test above
-                for ref mut gs in &mut self.game_slots {
-                    if gs.game_slot_id == *game_slot_id {
-                        for ref p in &self.players {
-                            if gs.player_ids.contains(&p.player_id) {
-                              players.push(p.player_name.clone());
-                            }
-                        }
-                        break;
+                let game_slot_id = &player.game_info.as_ref().unwrap().game_slot_id;  // unwrap ok because of test above
+                let gs = self.game_slots.get(&game_slot_id).unwrap();
+                // can we use filter AMEEN
+                for ref p in self.players.values() {
+                    if gs.player_ids.contains(&p.player_id) {
+                        players.push(p.name.clone());
                     }
                 }
                 return ResponseCode::PlayerList(players);
             },
-            RequestAction::ChatMessage(_)  => unimplemented!(),
+            RequestAction::ChatMessage(msg)  => {
+                let player = self.players.get(&player_id);
+                let player_in_game = player.is_some() && player.unwrap().game_info.is_some();
+
+                if !player_in_game {
+                    return ResponseCode::BadRequest(Some(format!("Player {} has not joined a game.", player_id)))
+                }
+
+                // User is in game, Server needs to broadcast this to GameSlot
+                let player = player.unwrap();
+                let slot_id = player.clone().game_info.unwrap().game_slot_id;
+
+                let opt_slot = self.game_slots.get_mut(&slot_id);
+
+                if opt_slot == None {
+                    return ResponseCode::BadRequest(Some(format!("No game found or player {} and slot id {}", player_id, slot_id)))
+                }
+
+                let gs = opt_slot.unwrap();
+                let ref mut messages: VecDeque<ServerChatMessage> = gs.messages;
+                let queue_size = messages.len();
+                gs.latest_seq_num += 1;
+
+                if queue_size >= MAX_NUM_CHAT_MESSAGES {
+                    messages.truncate(queue_size - MAX_NUM_CHAT_MESSAGES + 1);
```

Comment:
⚠️ This truncates from the back. We need it to truncate from the front! (Thanks @AaronM04 )

Reply:
Apparently truncate just calls `pop_back()` as many times as specified. Ended up writing something similar to `pop_front()` instead.